File: server/src/utils/tracking-summary.ts
/**
 * @fileoverview Utilities for building tracking data summaries.
 * Aggregates cookies, scripts, and network requests by domain
 * and prepares data for LLM analysis.
 */

import type {
  TrackedCookie,
  TrackedScript,
  NetworkRequest,
  StorageItem,
  DomainData,
  DomainBreakdown,
  TrackingSummary,
} from '../types.js'

/**
 * Group tracking data (cookies, scripts, network requests) by domain.
 * Creates a lookup table for analyzing which domains are tracking users.
 *
 * @param cookies - Array of tracked cookies
 * @param scripts - Array of detected scripts
 * @param networkRequests - Array of network requests
 * @returns Object mapping domain names to their tracking data
 */
function groupByDomain(
  cookies: TrackedCookie[],
  scripts: TrackedScript[],
  networkRequests: NetworkRequest[]
): Record<string, DomainData> {
  const domainData: Record<string, DomainData> = {}

  for (const cookie of cookies || []) {
    if (!domainData[cookie.domain]) {
      domainData[cookie.domain] = { cookies: [], scripts: [], networkRequests: [] }
    }
    domainData[cookie.domain].cookies.push(cookie)
  }

  for (const script of scripts || []) {
    if (!domainData[script.domain]) {
      domainData[script.domain] = { cookies: [], scripts: [], networkRequests: [] }
    }
    domainData[script.domain].scripts.push(script)
  }

  for (const request of networkRequests || []) {
    if (!domainData[request.domain]) {
      domainData[request.domain] = { cookies: [], scripts: [], networkRequests: [] }
    }
    domainData[request.domain].networkRequests.push(request)
  }

  return domainData
}

/**
 * Identify third-party domains relative to the analyzed URL.
 * Compares base domains to determine which are external trackers.
 *
 * @param domainData - Grouped tracking data by domain
 * @param analyzedUrl - The URL of the page being analyzed
 * @returns Array of domain names that are third-party
 */
function getThirdPartyDomains(domainData: Record<string, DomainData>, analyzedUrl: string): string[] {
  return Object.keys(domainData).filter((domain) => {
    try {
      const pageBaseDomain = new URL(analyzedUrl).hostname.split('.').slice(-2).join('.')
      const domainBase = domain.split('.').slice(-2).join('.')
      return pageBaseDomain !== domainBase
    } catch {
      return true
    }
  })
}

/**
 * Build a summary breakdown for each domain's tracking activity.
 * Extracts counts and types of tracking elements per domain.
 *
 * @param domainData - Grouped tracking data by domain
 * @returns Array of domain breakdown summaries
 */
function buildDomainBreakdown(domainData: Record<string, DomainData>): DomainBreakdown[] {
  return Object.entries(domainData).map(([domain, data]) => ({
    domain,
    cookieCount: data.cookies.length,
    cookieNames: data.cookies.map((c) => c.name),
    scriptCount: data.scripts.length,
    requestCount: data.networkRequests.length,
    requestTypes: [...new Set(data.networkRequests.map((r) => r.resourceType))],
  }))
}

/**
 * Build preview of storage items for analysis.
 * Truncates values to first 100 characters to reduce payload size.
 *
 * @param items - Array of storage items
 * @returns Array of objects with key and truncated value preview
 */
function buildStoragePreview(items: StorageItem[]): { key: string; valuePreview: string }[] {
  return (items || []).map((item) => ({
    key: item.key,
    valuePreview: item.value.substring(0, 100),
  }))
}

/**
 * Build a complete tracking summary for LLM privacy analysis.
 * Aggregates all tracking data into a structured format suitable for AI processing.
 *
 * @param cookies - Array of captured cookies
 * @param scripts - Array of detected scripts
 * @param networkRequests - Array of network requests
 * @param localStorage - Array of localStorage items
 * @param sessionStorage - Array of sessionStorage items
 * @param analyzedUrl - The URL of the page being analyzed
 * @returns Complete tracking summary with counts, domains, and breakdowns
 */
export function buildTrackingSummary(
  cookies: TrackedCookie[],
  scripts: TrackedScript[],
  networkRequests: NetworkRequest[],
  localStorage: StorageItem[],
  sessionStorage: StorageItem[],
  analyzedUrl: string
): TrackingSummary {
  const domainData = groupByDomain(cookies, scripts, networkRequests)

  return {
    analyzedUrl,
    totalCookies: cookies?.length || 0,
    totalScripts: scripts?.length || 0,
    totalNetworkRequests: networkRequests?.length || 0,
    localStorageItems: localStorage?.length || 0,
    sessionStorageItems: sessionStorage?.length || 0,
    thirdPartyDomains: getThirdPartyDomains(domainData, analyzedUrl),
    domainBreakdown: buildDomainBreakdown(domainData),
    localStorage: buildStoragePreview(localStorage),
    sessionStorage: buildStoragePreview(sessionStorage),
  }
}

File: server/src/utils/tracking_summary.py
"""
Utilities for building tracking data summaries.
Aggregates cookies, scripts, and network requests by domain
and prepares data for LLM analysis.
"""

from __future__ import annotations

import collections
from urllib import parse

from src.types import analysis, tracking_data


def _group_by_domain(
    cookies: list[tracking_data.TrackedCookie],
    scripts: list[tracking_data.TrackedScript],
    network_requests: list[tracking_data.NetworkRequest],
) -> dict[str, analysis.DomainData]:
    """Group tracking data (cookies, scripts, network requests) by domain."""
    domain_data: dict[str, analysis.DomainData] = collections.defaultdict(analysis.DomainData)

    for cookie in cookies:
        domain_data[cookie.domain].cookies.append(cookie)

    for script in scripts:
        domain_data[script.domain].scripts.append(script)

    for request in network_requests:
        domain_data[request.domain].network_requests.append(request)

    return domain_data


def _get_third_party_domains(
    domain_data: dict[str, analysis.DomainData], analyzed_url: str
) -> list[str]:
    """Identify third-party domains relative to the analyzed URL."""
    page_hostname = parse.urlparse(analyzed_url).hostname or ""
    page_base = ".".join(page_hostname.split(".")[-2:])

    results = []
    for domain in domain_data:
        domain_base = ".".join(domain.split(".")[-2:])
        if page_base != domain_base:
            results.append(domain)
    return results


def _build_domain_breakdown(domain_data: dict[str, analysis.DomainData]) -> list[analysis.DomainBreakdown]:
    """Build a summary breakdown for each domain's tracking activity."""
    return [
        analysis.DomainBreakdown(
            domain=domain,
            cookie_count=len(data.cookies),
            cookie_names=[c.name for c in data.cookies],
            script_count=len(data.scripts),
            request_count=len(data.network_requests),
            request_types=list({r.resource_type for r in data.network_requests}),
        )
        for domain, data in domain_data.items()
    ]


def _build_storage_preview(items: list[tracking_data.StorageItem]) -> list[dict[str, str]]:
    """Build preview of storage items for analysis."""
    return [{"key": item.key, "valuePreview": item.value[:100]} for item in items]


def build_tracking_summary(
    cookies: list[tracking_data.TrackedCookie],
    scripts: list[tracking_data.TrackedScript],
    network_requests: list[tracking_data.NetworkRequest],
    local_storage: list[tracking_data.StorageItem],
    session_storage: list[tracking_data.StorageItem],
    analyzed_url: str,
) -> analysis.TrackingSummary:
    """Build a complete tracking summary for LLM privacy analysis."""
    domain_data = _group_by_domain(cookies, scripts, network_requests)

    return analysis.TrackingSummary(
        analyzed_url=analyzed_url,
        total_cookies=len(cookies),
        total_scripts=len(scripts),
        total_network_requests=len(network_requests),
        local_storage_items=len(local_storage),
        session_storage_items=len(session_storage),
        third_party_domains=_get_third_party_domains(domain_data, analyzed_url),
        domain_breakdown=_build_domain_breakdown(domain_data),
        local_storage=_build_storage_preview(local_storage),
        session_storage=_build_storage_preview(session_storage),
    )
